linkerd/app/outbound/src/http/handle_proxy_error_headers.rs

```diff
@@ -0,0 +1,253 @@
+use futures::prelude::*;
+use linkerd_app_core::{
+    errors::respond::{L5D_PROXY_CONNECTION, L5D_PROXY_ERROR},
+    proxy::http::ClientHandle,
+    svc, tls,
+};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tracing::debug;
+
+#[derive(Copy, Clone, Debug)]
+pub struct Closable(pub bool);
+
+/// Close the accepted connection if the response from a peer proxy has the
+/// `l5d-proxy-connection: close` header. This means the peer proxy encountered
+/// an inbound connection error with its application and therefore the accepted
+/// connection should be torn down.
+#[derive(Clone, Debug)]
+pub struct NewHandleProxyErrorHeaders<X, N> {
+    inner: N,
+    extract: X,
+}
+
+#[derive(Clone, Debug)]
+pub struct HandleProxyErrorHeaders<S> {
+    inner: S,
+    closable: bool,
+}
+
+#[pin_project::pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F> {
+    #[pin]
+    inner: F,
+    client: ClientHandle,
+    closable: bool,
+}
+
+impl<X: Clone, N> NewHandleProxyErrorHeaders<X, N> {
+    fn new(extract: X, inner: N) -> Self {
+        Self { extract, inner }
+    }
+
+    pub fn layer_via(extract: X) -> impl svc::layer::Layer<N, Service = Self> + Clone {
+        svc::layer::mk(move |inner| Self::new(extract.clone(), inner))
+    }
+}
+
+impl<N> NewHandleProxyErrorHeaders<(), N> {
+    pub fn layer() -> impl svc::layer::Layer<N, Service = Self> + Clone {
+        Self::layer_via(())
+    }
+}
+
+impl<T, X, N> svc::NewService<T> for NewHandleProxyErrorHeaders<X, N>
+where
+    X: svc::ExtractParam<Closable, T>,
+    X: svc::ExtractParam<tls::ConditionalClientTls, T>,
```

Comment:
do we really want to use the same `ExtractParam` impl to extract both of these? i realize this probably doesn't actually matter as it looks like in practice, the `ExtractParam` impl is currently always `()`, and this is just here for future-proofing...

Reply:
Yeah, we can change it later if it needs to be changed.